### PR TITLE
cluster manager: change underlying call chain for drainConnections

### PR DESCRIPTION
**Pull request by junr03**
Commit Message: cluster manager - change underlying call chain for drainConnections
Additional Description: previously drainConnections used `ConnPool::startDrain` which does not guarantee that non-idle connections will not be used for new streams. This change leverages the logic in onHostHealthFailure to close idle connections and mark non-idle connections as draining so that new streams are forced onto new connections.
Risk Level: low - API usage is opt in
Testing: existing tests.

Signed-off-by: Jose Nino <jnino@lyft.com>

### CODE REVIEW

**Comment by junr03** [comment]
@mattklein123 I wanted to get this up asap to get your take. I felt that all the logic in onHostHealthFailure was what we wanted out of `ClusterManager::drainConnections()` (especially because that function does not charge any stats specific to host health failure) but I am the first to admit that it is not the cleanest code. I can move things a bit around, but given the subsequent cleanup I wanted to do after @RyanTheOptimist changes in the conn pool it felt silly, so I opted for a todo. lmk what you think.

re: testing. The existing test obviously passes, but I am still working on a test that more explicitly makes sure that a new connection is establish for a stream even if the previous connection was not idle. I had the idea to look for how onHostHealthFailure is tested this morning so I will push an update after my morning meetings.

File: source/common/upstream/cluster_manager_impl.cc [hunk 1]
```diff
@@ -1408,6 +1408,18 @@ void ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::drainConnP
   }
 }
 
+// TODO(junr03): clean up this usage and the usage above once ConnPoolImplBase::startDrain and
+// ConnPoolImplBase::drainConnections() get cleaned up. The code in onHostHealthFailure and the code
+// in ThreadLocalClusterManagerImpl::drainConnPools(const HostVector& hosts) is very similar and
+// can be merged in a similar fashion to the ConnPoolImplBase case.
```
Comment (mattklein123) [inline source/common/upstream/cluster_manager_impl.cc:1414]: I think you can just factor out the relevant part of both functions into a helper function that has a better name? I don't think we really need the TODOs here. We can just fix it later if needed.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks for fixing. Small comment.

/wait

File: source/common/upstream/cluster_manager_impl.cc (hunk 1 above)
Comment (junr03) [reply source/common/upstream/cluster_manager_impl.cc:1414]: alright. let me do that.

**Review by mattklein123 — APPROVED** [review]
Cool, thanks!

**Pull request merged**